File: app/src/main/java/tr/edu/iyte/quickreply/QuickReplyTile.java
package tr.edu.iyte.quickreply;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.drawable.Icon;
import android.service.quicksettings.Tile;
import android.service.quicksettings.TileService;
import android.telephony.PhoneStateListener;
import android.telephony.SmsManager;
import android.telephony.TelephonyManager;
import android.util.Log;

import java.lang.reflect.Method;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

import tr.edu.iyte.quickreply.activities.RequestPermissionActivity;
import tr.edu.iyte.quickreply.activities.SelectReplyActivity;
import tr.edu.iyte.quickreply.services.CallStopService;

public class QuickReplyTile extends TileService {
    public static class IncomingCallListener extends PhoneStateListener {
        private static final String TAG = "IncomingCallListener";

        private final Context c;

        public IncomingCallListener(Context c) {
            this.c = c;
        }

        @Override
        public void onCallStateChanged(int state, String incomingNumber) {
            super.onCallStateChanged(state, incomingNumber);
            if(state == TelephonyManager.CALL_STATE_RINGING) {
                endCall(c, incomingNumber);
                sendReply(currentReply ,incomingNumber);
                prefs.edit().putInt(SHARED_PREF_REPLY_COUNT_KEY, ++replyCount).apply();
            }
        }

        @SuppressWarnings("unchecked")
        private void endCall(Context context, String phoneNum) {
            try {
                TelephonyManager tm = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
                Class c = Class.forName(tm.getClass().getName());
                Method m = c.getDeclaredMethod("getITelephony");
                m.setAccessible(true);
                Object telephonyService = m.invoke(tm); // Get the internal ITelephony object
                c = Class.forName(telephonyService.getClass().getName()); // Get its class
                m = c.getDeclaredMethod("endCall"); // Get the "endCall()" method
                m.setAccessible(true); // Make it accessible
                m.invoke(telephonyService); // invoke endCall()
                Log.i(TAG, "Call blocked from <" + phoneNum + ">");
            } catch(Exception e) {
                Log.e(TAG, "couldn't end call", e);
            }
        }

        private void sendReply(String reply, String phoneNum) {
            SmsManager sms = SmsManager.getDefault();
            sms.sendTextMessage(phoneNum, null, reply, null, null);
            Log.i(TAG, "SMS <" + reply + "> sent to <" + phoneNum + ">");
        }
    }

    private static final String TAG = "QuickReplyTile";
    private static final String SHARED_PREF_KEY = "tr.edu.iyte.quickreply.sharedprefs";
    private static final String SHARED_PREF_PERMISSIONS_KEY = "tr.edu.iyte.quickreply.sharedprefs.perms";
    private static final String SHARED_PREF_REPLY_COUNT_KEY = "tr.edu.iyte.quickreply.sharedprefs.replycount";
    private static final String SHARED_PREF_REPLIES_KEY = "tr.edu.iyte.quickreply.sharedprefs.replies";

    private static final String DEFAULT_TITLE = "Quick Reply";
    private static final int DEFAULT_REPLY_COUNT = 0;
    private static final boolean DEFAULT_HAS_PERMISSIONS = false;
    private static final String DEFAULT_REPLY = "";
    private static final Set<String> DEFAULT_REPLIES = new HashSet<>();

    private static boolean hasPermissions = DEFAULT_HAS_PERMISSIONS;
    private static int replyCount = DEFAULT_REPLY_COUNT;
    private static Set<String> replies = DEFAULT_REPLIES;
    private static SharedPreferences prefs = null;
    private static String currentReply = DEFAULT_REPLY;

    @Override
    public void onTileAdded() {
        super.onTileAdded();
        if(DEFAULT_REPLIES.isEmpty()) {
            DEFAULT_REPLIES.add(getResources().getString(R.string.default_reply_1));
            DEFAULT_REPLIES.add(getResources().getString(R.string.default_reply_2));
            DEFAULT_REPLIES.add(getResources().getString(R.string.default_reply_3));
        }
        Log.d(TAG, "Tile added");
    }

    @Override
    public void onStartListening() {
        Log.i(TAG, "Started listening");
        if(prefs == null)
            prefs = getSharedPreferences(SHARED_PREF_KEY, MODE_PRIVATE);
        hasPermissions = prefs.getBoolean(SHARED_PREF_PERMISSIONS_KEY, DEFAULT_HAS_PERMISSIONS);
        replyCount = prefs.getInt(SHARED_PREF_REPLY_COUNT_KEY, DEFAULT_REPLY_COUNT);
        replies = prefs.getStringSet(SHARED_PREF_REPLIES_KEY, DEFAULT_REPLIES);

        updateTile();
    }

    @Override
    public void onStopListening() {
        Log.d(TAG, "Stopped listening");
    }

    @Override
    public void onTileRemoved() {
        Log.d(TAG, "Tile removed");
        replyCount = DEFAULT_REPLY_COUNT;
    }

    @Override
    public void onClick() {
        super.onClick();
        if(!hasPermissions) {
            startActivityAndCollapse(new Intent(this, RequestPermissionActivity.class));
            return;
        }

        if(!currentReply.isEmpty()) {
            reset();
            return;
        }

        startActivityAndCollapse(new Intent(this, SelectReplyActivity.class));
    }

    private void updateTile() {
        if(currentReply.isEmpty()) {
            defaultTile();
            return;
        }

        Tile t = getQsTile();
        t.setState(Tile.STATE_ACTIVE);
        t.setLabel(String.format(Locale.getDefault(), "(%d) %s", replyCount, currentReply));
        t.setIcon(Icon.createWithResource(this, R.drawable.ic_chat_bubble_black_24dp));
        t.updateTile();
    }

    private void defaultTile() {
        Tile t = getQsTile();
        t.setState(Tile.STATE_INACTIVE);
        t.setLabel(DEFAULT_TITLE);
        t.setIcon(Icon.createWithResource(this, R.drawable.ic_chat_bubble_outline_black_24dp));
        t.updateTile();
    }

    private void reset() {
        stopService(new Intent(this, CallStopService.class));
        defaultTile();
        currentReply = DEFAULT_REPLY;
        prefs.edit().putInt(SHARED_PREF_REPLY_COUNT_KEY, DEFAULT_REPLY_COUNT).apply();
        Log.i(TAG, "Reset");
    }

    public static void permissionsGranted() {
        hasPermissions = true;
        prefs.edit().putBoolean(SHARED_PREF_PERMISSIONS_KEY, hasPermissions).apply();
    }

    public static Set<String> getReplies() {
        return replies;
    }

    public static void selectReply(String reply) {
        currentReply = reply;
        Log.i(TAG, "Selected reply: " + reply);
    }

    public static boolean addReply(String reply) {
        replies.add(reply);
        prefs.edit().putStringSet(SHARED_PREF_REPLIES_KEY, replies).apply();
        Log.i(TAG, "Added new reply: " + reply);
        return replies.size() == 1;
    }

    public static void removeReply(String reply) {
        replies.remove(reply);
        prefs.edit().putStringSet(SHARED_PREF_REPLIES_KEY, replies).apply();
        Log.i(TAG, "Reply removed: " + reply);
    }

    public static boolean hasNoReply() {
        return replies.isEmpty();
    }
}

File: app/src/main/java/tr/edu/iyte/quickreply/QuickReplyTile.kt
package tr.edu.iyte.quickreply

import android.app.Activity
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.graphics.drawable.Icon
import android.os.IBinder
import android.service.quicksettings.Tile
import android.service.quicksettings.TileService
import android.telephony.PhoneStateListener
import android.telephony.SmsManager
import android.telephony.TelephonyManager
import org.jetbrains.anko.*
import tr.edu.iyte.quickreply.activities.ReplyRuleActivity
import tr.edu.iyte.quickreply.helper.database
import tr.edu.iyte.quickreply.helper.startActivityAndCollapse
import tr.edu.iyte.quickreply.helper.stopService
import tr.edu.iyte.quickreply.services.CallStopService

class QuickReplyTile : TileService(), AnkoLogger {
    override fun onBind(intent: Intent): IBinder? {
        ReplyManager.init(this)
        verbose("Bound")
        return super.onBind(intent)
    }

    override fun onStartListening() {
        super.onStartListening()
        info("Started listening")

        if(ReplyManager.currentReply.isEmpty())
            updateTile()
        else
            updateTile(state = Tile.STATE_ACTIVE,
                    label = "(${ReplyManager.replyCount}) ${ReplyManager.currentReply}",
                    icon = R.drawable.ic_chat_bubble_black_24dp)
    }

    override fun onStopListening() {
        info("Stopped listening")
        super.onStopListening()
    }

    override fun onClick() {
        super.onClick()
        info("Clicked")

        if (!ReplyManager.currentReply.isEmpty()) {
            reset()
            return
        }

        startActivityAndCollapse<ReplyRuleActivity>()
    }

    private fun updateTile(state: Int = Tile.STATE_INACTIVE,
                   label: String = "Quick Reply",
                   icon: Int = R.drawable.ic_chat_bubble_outline_black_24dp) {
        qsTile?.state = state
        qsTile?.label = label
        qsTile?.icon = Icon.createWithResource(this, icon)
        info("Updating tile with label: $label")
        qsTile?.updateTile()
    }

    private fun reset() {
        info("Resetting tile service")
        stopService<CallStopService>()
        updateTile()
        ReplyManager.replyCount = ReplyManager.DEFAULT_REPLY_COUNT
        ReplyManager.currentReply = ReplyManager.DEFAULT_REPLY
    }

    override fun onTileAdded() {
        super.onTileAdded()
        info("Tile added")
    }

    override fun onTileRemoved() {
        info("Tile removed")
        super.onTileRemoved()
    }

    companion object {
        internal const val SMS_DELIV_INTENT_REQUEST_CODE = 2 shl 6
        internal const val EXTRA_CALLER = "caller"
        internal const val EXTRA_TYPE = "type"
        internal const val EXTRA_SENT_ON = "sentOn"
        internal const val EXTRA_TYPE_PHONE = 0
        //internal const val EXTRA_TYPE_SMS = 1

        class SMSSentReceiver : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
                val isDelivered = resultCode == Activity.RESULT_OK

                context?.database?.addToHistory(
                        ReplyLog(callerNumber = intent!!.getStringExtra(EXTRA_CALLER),
                                reply = ReplyManager.currentReply,
                                type = if(intent.getIntExtra(EXTRA_TYPE, EXTRA_TYPE_PHONE) == EXTRA_TYPE_PHONE)
                                    ReplyLogType.CALL else ReplyLogType.SMS,
                                sentOn = intent.getLongExtra(EXTRA_SENT_ON, System.currentTimeMillis()),
                                isDelivered = isDelivered))
            }
        }

        class IncomingCallListener(val context: Context) : PhoneStateListener(), AnkoLogger {
            override fun onCallStateChanged(state: Int, incomingNumber: String?) {
                super.onCallStateChanged(state, incomingNumber)
                if (state == TelephonyManager.CALL_STATE_RINGING) {
                    endCall(context, incomingNumber!!)
                    sendReply(ReplyManager.currentReply, incomingNumber)
                    ReplyManager.replyCount++
                }
            }

            private fun endCall(context: Context, phoneNum: String) {
                try {
                    val tm = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
                    var c = Class.forName(tm.javaClass.name)
                    var m = c.getDeclaredMethod("getITelephony")
                    m.isAccessible = true
                    val telephonyService = m.invoke(tm) // Get the internal ITelephony object
                    c = Class.forName(telephonyService.javaClass.name) // Get its class
                    m = c.getDeclaredMethod("endCall") // Get the "endCall()" method
                    m.isAccessible = true // Make it accessible
                    m.invoke(telephonyService) // invoke endCall()
                    info("Call blocked from <$phoneNum>")
                } catch (e: Exception) {
                    error("Couldn't end call", e)
                }

            }

            private fun sendReply(reply: String, phoneNum: String) {
                val sms = SmsManager.getDefault()
                val receiverIntent = context.intentFor<SMSSentReceiver>(EXTRA_CALLER to phoneNum,
                        EXTRA_TYPE to EXTRA_TYPE_PHONE,
                        EXTRA_SENT_ON to System.currentTimeMillis())
                val sentIntent = PendingIntent.getBroadcast(context,
                        SMS_DELIV_INTENT_REQUEST_CODE,
                        receiverIntent,
                        PendingIntent.FLAG_ONE_SHOT)
                sms.sendTextMessage(phoneNum, null, reply, sentIntent, null)
                info("SMS <$reply> sent to <$phoneNum>")
            }
        }

        // TODO implement incoming sms listener
    }
}